language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -157,265 +260,231 @@ runCekM
+--  'FrameUnwrap': if v is a wrapped value w then returnCek w, else fail.
 returnCek
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> WithMemory Value uni -> CekM uni (Plain Term uni)
--- Instantiate all the free variable of the resulting term in case there are any.
-returnCek [] res = getVarEnv <&> \varEnv -> void $ dischargeVarEnv varEnv res
-returnCek (FrameTyInstArg ty : con) fun = instantiateEvaluate con ty fun
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
-    funVarEnv <- getVarEnv
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
-    argVarEnv <- getVarEnv
-    applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg : con) val =
-    returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
-    term             ->
-        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just term
+    => Context uni -> CekValue uni -> CekM uni (Plain Term uni)
+--- Instantiate all the free variable of the resulting term in case there are any.
+-- . ◅ V           ↦  [] V
+returnCek [] val = pure $ void $ dischargeCekValue val
+-- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
+returnCek (FrameTyInstArg ty : ctx) fun = instantiateEvaluate ctx ty fun
+-- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
+    withEnv argVarEnv $ computeCek (FrameApplyFun fun : ctx) arg
+-- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
+-- FIXME: add rule for VBuiltin once it's in the specification.
+returnCek (FrameApplyFun fun : ctx) arg = do
+    applyEvaluate ctx fun arg
+-- s , wrap A B _ ◅ V  ↦  s ◅ wrap A B V
+returnCek (FrameIWrap ex pat arg : ctx) val =
+    returnCek ctx $ VIWrap ex pat arg val
+-- s , unwrap _ ◅ wrap A B V  ↦  s ◅ V
+returnCek (FrameUnwrap : ctx) val =
+    case val of
+      VIWrap _ _ _ v -> returnCek ctx v
+      _              ->
+        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just val
+
+{- [Note: accumulating arguments].  The VBuiltin value contains lists of type and term
+ arguments which grow as new arguments are encountered.  In the code below We just add
+ new entries by appending to the end of the list: l -> l++[x].  This doesn't look
+ terrbily good, but we don't expect the lists to ever contain more than three or four
+ elements, so the cost is unlikely to be high.  We could accumulate lists in the normal
+ way and reverse them when required, but this is error-prone and reversal adds an extra
+ cost anyway.  We could also use something like Data.Sequence, but again we incur an
+ extra cost because we have to convert to a normal list when passing the arguments to
+ the constant application machinery.  If we really care we might want to convert the CAM
+ to use sequences instead of lists.
```

Comment:
I do think we should at least mention difference lists. They're almost as simple as regular ones, do not involve any substantial overhead, do not require us to reverse anything, etc.

Reply:
I forgot about that. I'll give it a try since I don't think I can actually face having to look at `args++[arg]`.